### PR TITLE
Encrypting user/email from cookie

### PR DESCRIPTION
**Pull request by costelmoraru**
## Description
This change does one thing:

if a cypher is provided, it encrypts also the email and user with the same logic and algorithm as the rest of the fields from the session state.

## Motivation and Context

This is the PR for the issue #60 .
Don't expose any information part of the cookie, information that can be easily retrieved if not encrypted.

## How Has This Been Tested?

All the test from session_state_test are passing.
Testing in our environments.

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

**Comment by costelmoraru** [comment]
@JoelSpeed Just pushed the code for the new json encoding of the cookie.

File: providers/session_state.go [hunk 1]
```diff
@@ -172,6 +185,20 @@ func DecodeSessionState(v string, c *cookie.Cipher) (*SessionState, error) {
 			User:  ss.User,
 		}
 	} else {
+		// Backward compatibility with using unecrypted Email
+		if ss.Email != "" {
+			decryptedEmail, errEmail := c.Decrypt(ss.Email)
```
Comment (JoelSpeed) [inline providers/session_state.go]: Is there any chance this `errEmail` could be `nil` whilst the original email wasn't encrypted? Just wondering if somehow `foo@example.com` could get garbled into `fnduoiafjd` by the decryption algorithm?

File: providers/session_state.go [hunk 2]
```diff
@@ -62,6 +62,19 @@ func (s *SessionState) EncodeSessionState(c *cookie.Cipher) (string, error) {
 	} else {
 		ss = *s
 		var err error
+		// Encrypt also Email and User when cipher is provided
```
Comment (JoelSpeed) [inline providers/session_state.go]: I'm not sure this comment is necessary, personally I would drop it but happy to leave if you feel strongly about it

File: providers/session_state.go (hunk 1 above)
Comment (costelmoraru) [reply providers/session_state.go]: No, is not. 
This scenario is covered by a unit test, the last test case from `TestDecodeSessionState`.
```
{
	SessionState: SessionState{
		Email:        "user@domain.com",
		User:         "just-user",
		AccessToken:  "token1234",
		IDToken:      "rawtoken1234",
		ExpiresOn:    e,
		RefreshToken: "refresh4321",
	},
	Encoded: fmt.Sprintf("email:user@domain.com user:just-user|I6s+ml+/MldBMgHIiC35BTKTh57skGX24w==|xojNdyyjB1HgYWh6XMtXY/Ph5eCVxa1cNsklJw==|%d|qEX0x6RmASxo4dhlBG6YuRs9Syn/e9sHu/+K", eUnix),
	Cipher:  c,
},
```
Trying to decrypt the email will make `errEmail` be `errorString{s:"failed to decrypt cookie value illegal base64 data at input byte 4"}`.

File: providers/session_state.go (hunk 2 above)
Comment (costelmoraru) [reply providers/session_state.go]: Yes, I guess from this point forward there is no different treatment in encrypting the email and user versus the other fields. Removing the comment.

File: providers/session_state.go (hunk 1 above)
Comment (JoelSpeed) [reply providers/session_state.go]: Excellent, thanks for verifying 😄

**Pull request merged**